### PR TITLE
Add filtering logic

### PR DESCRIPTION
**Pull request by dhaneshnm**
- Add filtering logic
- Refactor code organization.

TODO:
- Proper Styling for filter menu.
- Show the counts next to filter items.

### CODE REVIEW

File: src/subapps/search/components/TableHeightWrapper.tsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  resultTableHeightTestRef: React.MutableRefObject<HTMLDivElement>;
+  wrapperDOMProps: Bounds;
+}> = ({
+  wrapperHeightRef,
+  resultTableHeightTestRef,
+  wrapperDOMProps,
+  children,
+}) => {
+  return (
+    <div style={{ height: 'calc(100vh - 82px)' }}>
```
Comment (Stafeeva) [inline src/subapps/search/components/TableHeightWrapper.tsx:16]: Can we have css styles in a separate file please? :)

File: src/subapps/search/components/TableHeightWrapper.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        }}
+      >
+        <div
+          className="height-tester"
+          ref={wrapperHeightRef}
+          style={{ margin: '0' }}
+        >
+          <div
+            ref={resultTableHeightTestRef}
+            className={'result-table heightTest'}
```
Comment (Stafeeva) [inline src/subapps/search/components/TableHeightWrapper.tsx:30]:
```suggestion
            className='result-table heightTest'
```

File: src/subapps/search/components/TableHeightWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                },
+              ]}
+              pagination={{
+                position: ['topRight'],
+                showSizeChanger: true,
+              }}
+            ></Table>
+          </div>
+          <div
+            className={'result-table'}
```
Comment (Stafeeva) [inline src/subapps/search/components/TableHeightWrapper.tsx:55]:
```suggestion
            className='result-table'
```

File: src/subapps/search/hooks/useGlobalSearch.tsx [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    filterType: '',
+    filterTerm: '',
+  };
+  const [filterState, dispatchFilter] = React.useReducer(
+    filterReducer,
+    defaultFilter
+  );
+
+  const onFilterSubmit = (values: FilterState) => {
+    console.log('onFilterSubmit', values);
```
Comment (Stafeeva) [inline src/subapps/search/hooks/useGlobalSearch.tsx:190]:
```suggestion
    
```

File: src/subapps/search/hooks/useGlobalSearch.tsx [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+  const esQuery = React.useMemo(() => {
+    const baseQuery = constructQuery(query);
+    const withFilter = constructFilter(
+      baseQuery,
+      filterState.filters,
+      filterState.filterType,
+      filterState.filterTerm
+    );
+    const withPagination = addPagination(withFilter, from, size);
+    const yo = withPagination.build();
```
Comment (Stafeeva) [inline src/subapps/search/hooks/useGlobalSearch.tsx:217]: was it experimental? :)

File: src/subapps/search/hooks/useGlobalSearch.tsx [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    console.log(yo);
```
Comment (Stafeeva) [inline src/subapps/search/hooks/useGlobalSearch.tsx:218]:
```suggestion
    
```

File: src/subapps/search/hooks/useSearchPagination.tsx [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+  isInitialized: boolean;
+};
+
+/* ElasticSearch's index.max_result_window setting (default to 10k)
+determines how far we can page into results */
+export const ESMaxResultWindowSize = 10000;
+export const defaultPageSizeOptions = [10, 20, 50, 100];
+
+function useSearchPagination(query: any, nexus: NexusClient) {
+  const defaulpagination: SearchPagination = {
```
Comment (Stafeeva) [inline src/subapps/search/hooks/useSearchPagination.tsx:23]:
```suggestion
  const defaultPagination: SearchPagination = {
```

File: src/subapps/search/hooks/useSearchPagination.tsx [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+    currentPage: 1,
+    totalNumberOfResults: 0,
+    trueTotalNumberOfResults: 0,
+    pageSize: 0,
+    pageSizeOptions: [] as string[],
+    pageSizeFixed: false,
+    isInitialized: false,
+  };
+  const [pagination, setPagination] = React.useState<SearchPagination>(
+    defaulpagination
```
Comment (Stafeeva) [inline src/subapps/search/hooks/useSearchPagination.tsx:34]:
```suggestion
    defaultPagination
```

File: src/subapps/search/hooks/useSearchPagination.tsx [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+      }
+    },
+    [pagination]
+  );
+
+  const showTotal = React.useCallback(
+    (total: number, range: [number, number]) =>
+      pagination.trueTotalNumberOfResults <= ESMaxResultWindowSize ? (
+        <>
+          Showing {total.toLocaleString('en-US')}{' '}
```
Comment (Stafeeva) [inline src/subapps/search/hooks/useSearchPagination.tsx:78]: is it a way to convert a number to a string?

Comment (dhaneshnm) [reply src/subapps/search/hooks/useSearchPagination.tsx:78]: Pagination stuff and height stuff is a copy/refactor from existing code.  But will address your comments any way.

Comment (dhaneshnm) [reply src/subapps/search/hooks/useSearchPagination.tsx:78]: toLocaleString converts a number to string, but also makes sure that the string is dialect appropriate. for example, dates are written different in US vs Europe.

**Review by Stafeeva — APPROVED** [review]

**Pull request merged**